### PR TITLE
Make IOperation.ChildOperations public

### PR DESCRIPTION
**Pull request by 333fred**
Fixes https://github.com/dotnet/roslyn/issues/49475.

### CODE REVIEW

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler @CyrusNajmabadi @sharwell for review.

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IVariableDeclaration.cs
Comment (sharwell) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IVariableDeclaration.cs:615]: ❔ Fairly obvious question, but we do still have some tests for `Children` right? 😄

File: src/Compilers/Core/Portable/Operations/IOperation.Enumerable.cs
Comment (sharwell) [inline src/Compilers/Core/Portable/Operations/IOperation.Enumerable.cs:24]: ❔ Do value types nested in an interface cause problems for any common languages/versions? If I try to use it in a normal net472 app, it gives CS8370 about feature 'default interface implementation'. However, if I set `LangVersion` to 9 it does appear to execute without a runtime error.

File: src/Compilers/Core/Portable/Operations/IOperation.cs
Comment (sharwell) [inline src/Compilers/Core/Portable/Operations/IOperation.cs:50]:
```suggestion
        [Obsolete($"This API has performance penalties, please use {nameof(ChildOperations)} instead.", error: false)]
        [EditorBrowsable(EditorBrowsableState.Never)]
```

File: src/Compilers/Test/Core/CommonTestBase.cs
Comment (sharwell) [inline src/Compilers/Test/Core/CommonTestBase.cs:524]: 😕 Do we need `WhereNotNull()` here?

**Review by sharwell — APPROVED** [review]

**Review by cston — APPROVED** [review]

File: src/Compilers/Core/Portable/Operations/IOperation.Enumerable.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/IOperation.Enumerable.cs:24]: I did not realize that you couldn't do this before... I'll refactor this into `IOperationEnumerable`, I guess.

Comment (sharwell) [reply src/Compilers/Core/Portable/Operations/IOperation.Enumerable.cs:24]: I would not object to treating this as a point-in-time problem provided the following were met:

1. The framework team feels that this pattern reflects a long term "recommended design" for similar cases
2. There are no known cases that we need to support where this pattern causes runtime failures (complete inability to execute or use the feature)

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IVariableDeclaration.cs
Comment (333fred) [reply src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IVariableDeclaration.cs:615]: I'm not particularly worried about having dedicated tests for this API. It's deprecated and forwards directly to a well-tested version.

File: src/Compilers/Core/Portable/Operations/IOperation.Enumerable.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/IOperation.Enumerable.cs]: I don't think we want to recommend our consumers put themselves into an explicitly unsupported configuration.

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review.

File: src/Compilers/Core/Portable/Operations/IOperationEnumerable.cs
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Operations/IOperationEnumerable.cs:22]: calling this IOperationEumerable is... wonky to say the least.  Having the `I` pattern for a non-interface isn't something i've ever seen before.  I also don't think we do this with any of our other public enumerators.  I think we should follow this naming pattern:

```c#
ChildSyntaxList ChildNodesAndTokens()
```

So in this case, it would be `ChildOperationList ChildOperations`.

**Review by RikkiGibson — APPROVED** [review]
LGTM modulo Cyrus's comment https://github.com/dotnet/roslyn/pull/52923#discussion_r620814640

File: src/Compilers/Core/Portable/Operations/IOperationEnumerable.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/IOperationEnumerable.cs]: Renamed.

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
New name looks great.  Thanks!

**Comment by 333fred** [comment]
> New name looks great. Thanks!

@cston pointed out that it's not actually a list, it's an enumerable. I'm going to rename to `ChildOperationEnumerable`.

**Comment by sharwell** [comment]
@333fred I like the name `List`, plus it's not out of the question that we add `IReadOnlyList<IOperation>` to it in the future.

**Comment by RikkiGibson** [comment]
It feels like as designed, this type doesn't know the length of the sequence, and doesn't provide O(1) random access to items in the sequence. So I don't object to calling it a ChildOperationEnumerable.

**Comment by CyrusNajmabadi** [comment]
I think 'List' is fine, esp as i would imagine that we could add random access slot lookup in the future.

File: src/Compilers/Core/Portable/Operations/ChildOperationList.cs
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Operations/ChildOperationList.cs:16]: can you keep the file name in sync?

File: src/Compilers/Core/Portable/Operations/ChildOperationList.cs
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Operations/ChildOperationList.cs:44]: can you seed with the count?  or is it unknown?

File: src/Compilers/Core/Portable/Operations/ChildOperationList.cs
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Operations/ChildOperationList.cs:71]: based on the impl here, it appears as if Operations do understand slots/indexes.  so i think we should expose ChildOepration list and add more to it later on demand.  for example, things like First()/Last() and other helpers on ChlidSyntaxList have proven *very* helpful.  Thank you for coming to my Ted talk.

File: src/Compilers/Core/Portable/Operations/IOperation.cs
Comment (cston) [inline src/Compilers/Core/Portable/Operations/IOperation.cs:50]: >[Obsolete($"This API has performance penalties, please use {nameof(ChildOperations)} instead.", error: false)] [](http://example.com/codeflow?start=8&length=110)

The implementation simply delegates to `ChildOperations` so is boxing the only perf issue? (Depending on the callsite, boxing might be required.) 

If so, consider removing the `[Obsolete]` attribute or at least remove the warning.

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Operations/IOperation.cs:50]: This warning is generally waht we use for perf obsoletion.  We really do not want people using teh old system and incurring double allocations when this is availble and should be suitable for all hte same purposes.

File: src/Compilers/Core/Portable/Operations/ChildOperationList.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ChildOperationList.cs:44]: This is entirely unchanged from existing implementation. But yes, count is unknown unless we enumerate the list.

File: src/Compilers/Core/Portable/Operations/IOperation.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/IOperation.cs:50]: It's not just allocations here, it's also additional virtual calls in every foreach. Obsoleteing this is intentional.

File: src/Compilers/Core/Portable/Operations/IOperation.ChildCollection.cs
Comment (RikkiGibson) [inline src/Compilers/Core/Portable/Operations/IOperation.ChildCollection.cs:21]: nit: should this comment be changed to `default(ChildCollection)`?

File: src/Compilers/Core/Portable/Operations/IOperation.ChildCollection.Reversed.cs
Comment (Youssef1313) [inline src/Compilers/Core/Portable/Operations/IOperation.ChildCollection.Reversed.cs:26]: Note: `NonDefaultableAttribute` is used by an internal analyzer. This attribute has no effect for consumers outside Roslyn. Consider doc'ing that consumers shouldn't use `default(IOperation.ChildCollection.Reversed)` or something?

File: src/Compilers/Core/Portable/Operations/IOperation.ChildCollection.Reversed.cs
Comment (Youssef1313) [inline src/Compilers/Core/Portable/Operations/IOperation.ChildCollection.Reversed.cs:51]: Assert on null operation, or maybe even throw?

File: src/Compilers/Core/Portable/Operations/Operation.cs
Comment (Youssef1313) [inline src/Compilers/Core/Portable/Operations/Operation.cs:12]: nit: Consider moving these to the top.

The current usings are:

```
System.*
Microsoft.*
System.*
```

File: src/Compilers/Core/Portable/Operations/IOperation.ChildCollection.Reversed.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/IOperation.ChildCollection.Reversed.cs:52]: `enumerator.Current` will null assert and null ref, which is what it is documented to do for `default`, so I think this is fine.

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for reviews please.

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler please review.

**Review by RikkiGibson — DISMISSED** [review]

File: src/Compilers/Core/Portable/Operations/IOperation.ChildCollection.cs
Comment (cston) [inline src/Compilers/Core/Portable/Operations/IOperation.ChildCollection.cs:75]: >Reverse() [](http://example.com/codeflow?start=28&length=9)

Where is `Reverse()` used?

Comment (333fred) [reply src/Compilers/Core/Portable/Operations/IOperation.ChildCollection.cs:75]: API review asked for the APIs we expose for the syntax version of this list to be added, including `Reverse()`. In this particular PR, we use it for implementing `Last()`, without having to walk the entire collection.

Comment (cston) [reply src/Compilers/Core/Portable/Operations/IOperation.ChildCollection.cs:75]: Is there a test for `Reverse()`? If we're testing the forward iterators of each `IOperation` implementation, please consider testing that `Reverse()` returns the reverse order.

File: src/Compilers/Core/Portable/Operations/IOperation.ChildCollection.cs
Comment (cston) [inline src/Compilers/Core/Portable/Operations/IOperation.ChildCollection.cs:46]: >when !enumerator.MoveNext() [](http://example.com/codeflow?start=31&length=27)

It looks like the `default` branch depends on the execution order of the `case` and `when` clauses, which seems undefined. Please consider checking `Count == 0` explicitly instead, and move the `Enumerator` instance into the `default` branch.

File: src/Compilers/Core/Portable/Operations/IOperation.ChildCollection.Reversed.cs
Comment (cston) [inline src/Compilers/Core/Portable/Operations/IOperation.ChildCollection.Reversed.cs:49]: >var _ when !enumerator.MoveNext() [](http://example.com/codeflow?start=29&length=33)

The `default` branch depends on the execution order of `case` and `when` clauses. Please consider checking `Count == 0` here instead.

File: src/Compilers/Core/Portable/Operations/IOperation.ChildCollection.Reversed.cs
Comment (cston) [inline src/Compilers/Core/Portable/Operations/IOperation.ChildCollection.Reversed.cs:100]: >{ } [](http://example.com/codeflow?start=45&length=3)

Please implement and test.

File: src/Compilers/Core/Portable/Operations/IOperation.ChildCollection.cs
Comment (RikkiGibson) [reply src/Compilers/Core/Portable/Operations/IOperation.ChildCollection.cs:46]: Could you please clarify this? It seems like since `var _` always matches, it is well-defined that the 'default' label here is only reachable if `!enumerator.MoveNext()` is evaluated and returned `false`.

If you object as a matter of style, that's understandable, but I personally don't have an issue with it.

Comment (cston) [reply src/Compilers/Core/Portable/Operations/IOperation.ChildCollection.cs:46]: Does the compiler guarantee that a `when` clause is executed at most once? And is it good practice to have a `when` clause that has side-effects, such as modifying `enumerator`?

We should avoid these questions entirely, and simply refactor the code. For instance, either use `when Count == 0:` here or combine this `case` and the `default`:
```csharp
default:
    if (Count == 0)
        return ImmutableArray<IOperation>.Empty;
    var builder = ArrayBuilder<IOperation>.GetInstance(Count);
    ...
    return builder.ToImmutableAndFree();
```

Comment (RikkiGibson) [reply src/Compilers/Core/Portable/Operations/IOperation.ChildCollection.cs:46]: > Does the compiler guarantee that a when clause is executed at most once?

That is how the implementation behaves, and I would be very surprised if we ever changed that intentionally. The spec is very light on this, which would be good to address in another venue. There is a considerable compatibility burden for what we have shipped regardless of what we have declared to be implementation-defined.

I just don't want us to treat these constructs as something which may start behaving in an incompatible manner in next year's compiler. At this point any breaking change that we make to patterns would have to be something people are very unlikely to be depending on.

> We should avoid these questions entirely, and simply refactor the code.

I would like for us to have an "official" answer for the above question in a separate venue from this PR. As far as our actual development practices go, what you've suggested works for me. I can't argue against the claim that `enumerator.MoveNext()` in a when clause is harder to follow.

File: src/Compilers/Test/Core/CommonTestBase.cs
Comment (cston) [inline src/Compilers/Test/Core/CommonTestBase.cs:526]: >child is null [](http://example.com/codeflow?start=20&length=13)

When is `child is null`?

File: src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.cs
Comment (cston) [inline src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.cs:848]: >psuedocode [](http://example.com/codeflow?start=143&length=10)

Typo.

**Comment by jcouv** [comment]
@333fred I'm marking this PR as draft for now to make room in our review queue. Ping when ready for another look. Thanks

File: src/Compilers/Test/Core/CommonTestBase.cs
Comment (333fred) [reply src/Compilers/Test/Core/CommonTestBase.cs:526]: Theoretically never, this is likely some old code from the original IOperation implementation. Should be safe to remove.

**Comment by 333fred** [comment]
@cston please take another look.

File: src/Compilers/Core/Portable/Operations/IOperation.OperationList.cs
Comment (cston) [inline src/Compilers/Core/Portable/Operations/IOperation.OperationList.cs:21]: >ChildCollection [](http://example.com/codeflow?start=26&length=15)

`OperationList`

Comment (cston) [inline src/Compilers/Core/Portable/Operations/IOperation.OperationList.cs:21]: I think this sentence can be omitted since this is typical for `struct` types.

File: src/Compilers/Core/Portable/Operations/IOperation.OperationList.cs
Comment (cston) [inline src/Compilers/Core/Portable/Operations/IOperation.OperationList.cs:102]: Remove?

File: src/Compilers/Core/Portable/Operations/IOperation.OperationList.cs
Comment (cston) [inline src/Compilers/Core/Portable/Operations/IOperation.OperationList.cs:136]: >void IDisposable.Dispose() { } [](http://example.com/codeflow?start=16&length=30)

Does this mean that uses of `Enumerator` will typically require a `try/finally` block and require boxing the instance as `IDisposable`?

File: src/Compilers/Core/Portable/Operations/IOperation.OperationList.cs
Comment (cston) [inline src/Compilers/Core/Portable/Operations/IOperation.OperationList.cs:12]: >using [](http://example.com/codeflow?start=0&length=5)

Consider sorting and removing unused usings.

File: src/Compilers/Core/Portable/Operations/IOperation.OperationList.Reversed.cs
Comment (cston) [inline src/Compilers/Core/Portable/Operations/IOperation.OperationList.Reversed.cs:46]: >children [](http://example.com/codeflow?start=35&length=8)

Is this collection reversed? Should it be?

File: src/Compilers/Core/Portable/Operations/IOperation.OperationList.Reversed.cs
Comment (cston) [inline src/Compilers/Core/Portable/Operations/IOperation.OperationList.Reversed.cs:40]: >ToImmutableArray [](http://example.com/codeflow?start=50&length=16)

Is this method necessary?

File: src/Compilers/Core/Portable/Operations/IOperation.OperationList.cs
Comment (cston) [inline src/Compilers/Core/Portable/Operations/IOperation.OperationList.cs:117]: >_currentSlot >= 0 && _currentIndex >= 0 [](http://example.com/codeflow?start=59&length=39)

`Current` is a `public` property so I don't think we can assert this. For instance, this will fail with `new Enumerator(operation).Current`.

File: src/Compilers/Core/Portable/Operations/IOperation.OperationList.Reversed.cs
Comment (cston) [inline src/Compilers/Core/Portable/Operations/IOperation.OperationList.Reversed.cs:87]: >_currentSlot is >= 0 and not int.MaxValue && _currentIndex is >= 0 and not int.MaxValue [](http://example.com/codeflow?start=63&length=87)

`Current` is a `public` property so we probably cannot assert this.

File: src/Compilers/Core/Portable/Operations/IOperation.OperationList.Reversed.cs
Comment (cston) [inline src/Compilers/Core/Portable/Operations/IOperation.OperationList.Reversed.cs:106]: >void IDisposable.Dispose() { } [](http://example.com/codeflow?start=20&length=30)

Does this require `try/finally` and boxing?

File: src/Compilers/Core/Portable/Operations/OperationNodes.cs
Comment (cston) [inline src/Compilers/Core/Portable/Operations/OperationNodes.cs:380]: Consider using `if (!Arguments.IsEmpty) ...` to match the style of the other implementations.

File: src/Compilers/Core/Portable/Operations/OperationNodes.cs
Comment (cston) [inline src/Compilers/Core/Portable/Operations/OperationNodes.cs:466]: Consider using `if (!Arguments.IsEmpty) ...` to match the style of the other implementations.

File: src/Compilers/Core/Portable/Operations/IOperation.OperationList.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/IOperation.OperationList.cs:117]: That is not a constructor you can call (publicly). The only way to obtain an instance of `Enumerator` is use `default`, or get one from an `Enumerable`. In either case, we should assert so that, in debug code, you know you're doing something wrong.

File: src/Compilers/Core/Portable/Operations/IOperation.OperationList.Reversed.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/IOperation.OperationList.Reversed.cs:87]: Same comment as on the other `Current` implementation.

File: src/Compilers/Core/Portable/Operations/IOperation.OperationList.Reversed.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/IOperation.OperationList.Reversed.cs:40]: It is part of the approved public API, so I'm keeping it.

File: src/Compilers/Core/Portable/Operations/IOperation.OperationList.cs
Comment (cston) [inline src/Compilers/Core/Portable/Operations/IOperation.OperationList.cs:144]: >class [](http://example.com/codeflow?start=20&length=5)

`sealed`?

File: src/Compilers/Core/Portable/Operations/IOperation.OperationList.Reversed.cs
Comment (cston) [inline src/Compilers/Core/Portable/Operations/IOperation.OperationList.Reversed.cs:124]: >class [](http://example.com/codeflow?start=24&length=5)

`sealed`?

**Review by cston — APPROVED** [review]

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review.

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review please.

File: src/Compilers/Core/Portable/Operations/IOperation.OperationList.cs
Comment (cston) [reply src/Compilers/Core/Portable/Operations/IOperation.OperationList.cs]: Actually, the compiler emits a `constrained. callvirt` to `IDisposable.Dispose()` which should avoid boxing. Sorry for the confusion.

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review please.

**Comment by 333fred** [comment]
@sharwell @CyrusNajmabadi could you please take another look, particularly at my update to `GetBestTypeByMetadataName`?

File: src/Compilers/Core/Portable/Operations/OperationNodes.cs
Comment (jaredpar) [inline src/Compilers/Core/Portable/Operations/OperationNodes.cs:315]: The `MoveNextReversed` implementations differ on how they handle invalid `previousSlot` values. Some throw, some return `false`. Feel like the implementations should be consistent.

File: src/Compilers/Core/Portable/Operations/OperationNodes.cs
Comment (jaredpar) [inline src/Compilers/Core/Portable/Operations/OperationNodes.cs:60]: This one never throws, just returns `false`

**Comment by 333fred** [comment]
@jaredpar please take another look, I fixed the two inconsistent MoveNextReversed methods.

**Review by jaredpar — APPROVED** [review]

**Pull request merged**
